builtins/web/blob.cpp

```diff
@@ -0,0 +1,703 @@
+      return false;
+    }
+
+    auto endings_str = JS::ToString(cx, endings_val);
+    if (!JS_StringEqualsLiteral(cx, endings_str, "transparent", &is_transparent) ||
+        !JS_StringEqualsLiteral(cx, endings_str, "native", &is_native)) {
+      return false;
+    }
+
+    if (is_transparent || is_native) {
+      auto endings = is_native ? LineEndings::Native : LineEndings::Transparent;
+      SetReservedSlot(self, static_cast<uint32_t>(Slots::Endings), JS::Int32Value(endings));
+    }
+  }
+
+  if (has_type) {
+    JS::RootedValue type(cx);
+    if (!JS_GetProperty(cx, opts, "type", &type)) {
+      return false;
+    }
+
+    auto type_str = normalize_type(cx, type);
+    if (!type_str) {
+      return false;
+    }
+    SetReservedSlot(self, static_cast<uint32_t>(Slots::Type), PrivateValue(type_str));
+  }
+
+  return true;
+}
+
+JSObject *Blob::create(JSContext *cx, std::unique_ptr<Blob::ByteBuffer> data, JSString *type) {
+  JSObject *self = JS_NewObjectWithGivenProto(cx, &class_, proto_obj);
+  if (!self) {
+    return nullptr;
+  }
+
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Data), PrivateValue(data.release()));
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Type), PrivateValue(type));
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Endings), JS::Int32Value(LineEndings::Transparent));
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Readers), PrivateValue(new ReadersMap));
+  return self;
+}
+
+bool Blob::constructor(JSContext *cx, unsigned argc, JS::Value *vp) {
+  CTOR_HEADER("Blob", 0);
+
+  RootedValue blobParts(cx, args.get(0));
+  RootedValue opts(cx, args.get(1));
+  RootedObject self(cx, JS_NewObjectForConstructor(cx, &class_, args));
+
+  if (!self) {
+    return false;
+  }
+
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Data), PrivateValue(new std::vector<uint8_t>()));
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Type), PrivateValue(JS_GetEmptyString(cx)));
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Endings), JS::Int32Value(LineEndings::Transparent));
+  SetReservedSlot(self, static_cast<uint32_t>(Slots::Readers), PrivateValue(new ReadersMap));
+
+  // walk the blob parts and write it to concatenated buffer
+  if (blobParts.isNull()) {
+    return api::throw_error(cx, api::Errors::TypeError, "Blob.constructor", "blobParts", "be an object");
+  }
+
+  if (!blobParts.isUndefined() && !init_blob_parts(cx, self, blobParts)) {
+    return false;
+  }
+
+  if (!opts.isNullOrUndefined() && !init_options(cx, self, opts)) {
+    return false;
+  }
+
+  args.rval().setObject(*self);
+  return true;
+}
+
+bool Blob::init_class(JSContext *cx, JS::HandleObject global) {
+  return init_class_impl(cx, global);
+}
+
+void Blob::finalize(JS::GCContext *gcx, JSObject *self) {
+  auto blob = Blob::blob(self);
+  if (blob) {
+    free(blob);
+  }
+
+  auto readers = Blob::readers(self);
+  if (readers) {
+    free(readers);
```

Comment:
IIUC, a `BlobReader` doesn't own the `span` it holds, but instead is a view onto the `Blob`'s underlying buffer, right? If so, this seems correct, but would be good to document as part of the `BlobReader` definition.

Reply:
The blob buffer is something I plan to change, but in a follow-up PR. I already have a sketch for this in the Request#blob API work. Specifically, I’d like to switch from `std::vector` to `mozilla::Vector`. The reason is that mozilla::Vector provides optimizations that aren’t possible with `std::vector`, such as building a vector from raw parts, which allows creating a blob without copying buffers but simply by transferring the ownership.